pkg/workflows/sdk/builder.go

```diff
@@ -119,6 +138,16 @@ func NewWorkflowSpecFactory(
 // AddTo is meant to be called by generated code
 func (step *Step[O]) AddTo(w *WorkflowSpecFactory) CapDefinition[O] {
 	stepDefinition := step.Definition
+
+	if w.serialMode {
+		// ensure we depend on each previous step
+		for _, prevRef := range w.prevRefs {
+			if !stepDefinition.Inputs.HasRef(prevRef) {
+				stepDefinition.Condition = fmt.Sprintf("$(%s.success)", prevRef)
+			}
+		}
```

Comment:
Since it's serial you could just depend on the step right before it

Reply:
That is true if serial mode is left from the start, but if they just flipped it on, then we have to wait for any async nodes that were added since flipping off serial.